Add colored console output

poelzi opened the pull request:
**TODO:**
- [ ] Rebase after merging #3204 

Colorize log level, thread-name and category of console output.
Supports the NO_COLOR environment for disabling color output.

<img width="1101" alt="Screenshot_20201021_220346" src="https://user-images.githubusercontent.com/66107/96776456-58059280-13e9-11eb-8c6c-3f46188b7099.png">

Errors, warnings are clearly visible. Controllers and debug output is also clearly distinguished.

Holzhaus commented:
These ANSI escape codes are only supported on Unix. Let's find out that broke `QT_MESSAGE_PATTERN` support and just use that instead.

Holzhaus commented:
> These ANSI escape codes are only supported on Unix. Let's find out that broke `QT_MESSAGE_PATTERN` support and just use that instead.

Found the issue. Just apply this patch:

```diff
diff --git a/src/util/logging.cpp b/src/util/logging.cpp
index c967947789..047f9a4902 100644
--- a/src/util/logging.cpp
+++ b/src/util/logging.cpp
@@ -195,24 +195,9 @@ void handleMessage(
         }
     }

-    QByteArray ba;
-    ba.reserve(baSize);
-
-    ba.append(levelName);
-    ba.append(" [");
-    ba.append(threadName, threadName.size());
-    if (categoryName) {
-        ba.append("] ");
-        ba.append(categoryName, categoryName_len);
-        ba.append(": ");
-    } else {
-        ba.append("]: ");
-    }
-    ba.append(input8Bit, input8Bit.size());
-    ba.append('\n');
-
-    // Verify that the reserved size matches the actual size
-    DEBUG_ASSERT(ba.size() == baSize);
+    QString logMessage = qFormatLogMessage(type, context, input);
+    QByteArray ba = logMessage.toLocal8Bit();
+    ba.append("\n");

     if (isDebugAssert) {
         if (s_debugAssertBreak) {
```

Then you can set `QT_MESSAGE_PATTERN` however you want, e.g.:

```
export QT_MESSAGE_PATTERN="`echo -e "\033[32m%{time h:mm:ss.zzz}%{if-category}\033[32m %{category}:%{endif} %{if-debug}\033[34m%{function}%{endif}%{if-warning}\033[31m%{backtrace depth=3}%{endif}%{if-critical}\033[31m%{backtrace depth=3}%{endif}%{if-fatal}\033[31m%{backtrace depth=3}%{endif}\033[0m %{message}"`"
```

poelzi commented:
newer consoles of windows 10, alternative consoles, macosx support ansi colors. only old windows cmd does not support it.

Holzhaus commented:
> newer consoles of windows 10, alternative consoles, macosx support ansi colors. only old windows cmd does not support it.

OK, but why choose a more complex and way less powerful solution than what I posted above?

poelzi commented:
<img width="1101" alt="Screenshot_20201023_010411" src="https://user-images.githubusercontent.com/66107/96938865-b99e2d80-14cb-11eb-896f-aa1915b01685.png">

uklotzde reviewed (changes requested):
Let's first merge #3204 before introducing any custom extensions.

Be-ing commented:
#3204 has been merged. What is that status of this now?

poelzi commented:
@Be-ing I don't know. The problem wtih #3204 is, that it now pollutes the log file with ansi codes if I set there something that looks nice. I added it to my nix-shell environment because I know that I will not look at those files and only use the log file.

So, I'm somehow inclined to simply close this, otherwise we would need some ansi sequence remover before writing to the log file.

poelzi commented:
We could use https://code.woboq.org/qt5/qt-creator/src/libs/utils/ansiescapecodehandler.cpp.html to create the logfile version

poelzi commented:
This one is better: https://stackoverflow.com/questions/43743242/qregularexpression-to-strip-ansi-control-codes

uklotzde commented:
I am against adding more code to a performance critical path. Instead you could color your log file afterwards by applying regular expressions. This is nothing that needs to be added to production code where it runs for all users if only a few developers are actually using such a feature.

Only exception: Hidden behind a feature flag that is disabled by default. But then we still have to maintain the addtional code.

Holzhaus commented:
The proper way would be to format log file messages always with the standard format, and only use the custom pattern to the terminal output. I don't know if that is possible though.

uklotzde commented:
> The proper way would be to format log file messages always with the standard format, and only use the custom pattern to the terminal output. I don't know if that is possible though.

Exactly, it should be additive and optional, not subtractive and mandatory.

poelzi commented:
I misread the code. the formatting is only used for the terminal

poelzi committed:
Add option for colored output

Default is --colors auto which detects if user runs from terminal and
no NO_COLOR env is set.
Argument always forces colors, never prevents them

poelzi commented:
This is a new minimal version of it. I'm using this format now since the pull request and came to like it really much.
Looks like this:
<img width="1226" alt="Screenshot_20201214_000133" src="https://user-images.githubusercontent.com/66107/102026711-a9a21c00-3d9f-11eb-861d-1bd6ac1d2d3f.png">

poelzi commented:
Would be cool if the scripting engine could determine which js function triggered the debug() and would fill the variable, but otherwise I'm super happy with that.

Holzhaus commented on src/util/cmdlineargs.cpp:
```diff
@@ -152,6 +161,8 @@ void CmdlineArgs::printUsage() {
                         and spinning vinyl widgets. Try this option if\n\
                         Mixxx is crashing on startup.\n\
 \n\
+--noColors              Don't print colors on the console output.\n\
```
Rather than a `--noColors` flag we should use `--color=never` or `always` or `auto` (`auto` checks if the process is attached to a TTY) like `ls`.

poelzi replied:
good idea, done

daschuer reviewed (changes requested):
```
15:22:34.647 [Main] warning  QCheckBox(0x55fdbf655b00, name="LibraryBPMButton")  does not have a property named  "alignment"
              warning [Main] QCheckBox(0x55afcafb1fa0, name="LibraryBPMButton")  does not have a property named  "alignment"
```
This swaps type and thread name. This should not happen. 
This also adds a time stamp which is not related to color. 

Not sure about the default value --color=auto. Mixxx is no console application so it should not waste CPU time to beautify the console output. However it is now only if one starts Mixxx from a console and makes Warnings more visible. 

Maybe a good trade off. What doe other think?  

In the same scene we can add a bash-completion snippet. :-P

poelzi committed:
Merge branch 'main' of https://github.com/mixxxdj/mixxx into color_logging2

poelzi commented:
@daschuer changed the layout to reassemble the normal layout

daschuer commented:
For the future: Please do not rebase an active PR. Else we need to review the PR from the beginning instead of only the fix.

daschuer reviewed (approved):
LGTM, Thank you.

daschuer commented:
@uklotzde What do you think now?

Holzhaus commented:
I don't see anything in the diff about the log file. Generally this looks good, but the log file should always be written without colors. Is that the case? I'm short on time atm, I'll have a look asap.

poelzi commented:
@Holzhaus the file log is always without colors. The default is no colors and only if the stdout is detected as a tty the console output is colorized.

uklotzde reviewed (approved):
Works well and is helpful for debugging. Thank you! LGTM

Be-ing commented:
Thanks for working on this. Mixxx's console output is easier to read now.
